File: src/cpp/cdm/patient/actions/SEPneumoniaExacerbation.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "cdm/CommonDefs.h"
#include "cdm/patient/actions/SEPneumoniaExacerbation.h"
#include "cdm/properties/SEScalar0To1.h"
#include "cdm/io/protobuf/PBPatientActions.h"

SEPneumoniaExacerbation::SEPneumoniaExacerbation(Logger* logger) : SEPatientAction(logger)
{

}

SEPneumoniaExacerbation::~SEPneumoniaExacerbation()
{
  DELETE_MAP_SECOND(m_Severities);
}

void SEPneumoniaExacerbation::Clear()
{
  SEPatientAction::Clear();
  for (auto itr : m_Severities)
    if (itr.second != nullptr)
      itr.second->Invalidate();
}

void SEPneumoniaExacerbation::Copy(const SEPneumoniaExacerbation& src, bool /*preserveState*/)
{
  //if(preserveState) // Cache any state before copy,
  PBPatientAction::Copy(src, *this);
  //if(preserveState) // Put back any state
}

bool SEPneumoniaExacerbation::IsValid() const
{
  return HasSeverity();
}

bool SEPneumoniaExacerbation::IsActive() const
{
  if (!SEPatientAction::IsActive())
    return false;
  if (!HasSeverity())
    return false;
  return true;
}
void SEPneumoniaExacerbation::Deactivate()
{
  SEPatientAction::Deactivate();
  Clear();//No stateful properties
}

const SEScalar* SEPneumoniaExacerbation::GetScalar(const std::string& name)
{
  return nullptr;
}

bool SEPneumoniaExacerbation::HasSeverity() const
{
  for (auto itr : m_Severities)
    if (itr.second != nullptr && itr.second->IsPositive())
      return true;
  return false;
}
LungImpairmentMap& SEPneumoniaExacerbation::GetSeverities()
{
  return m_Severities;
}
const LungImpairmentMap& SEPneumoniaExacerbation::GetSeverities() const
{
  return m_Severities;
}

bool SEPneumoniaExacerbation::HasSeverity(eLungCompartment cmpt) const
{
  auto s = m_Severities.find(cmpt);
  if (s == m_Severities.end())
    return false;
  if (s->second == nullptr)
    return false;
  return s->second->IsValid();
}
SEScalar0To1& SEPneumoniaExacerbation::GetSeverity(eLungCompartment cmpt)
{
  SEScalar0To1* s = m_Severities[cmpt];
  if (s == nullptr)
  {
    s = new SEScalar0To1();
    m_Severities[cmpt] = s;
  }
  return *s;
}
const SEScalar0To1* SEPneumoniaExacerbation::GetSeverity(eLungCompartment cmpt) const
{
  auto s = m_Severities.find(cmpt);
  if (s == m_Severities.end())
    return nullptr;
  return s->second;
}

File: src/csharp/pulse/cdm/patient/actions/SEPneumoniaExacerbation.cs
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

using System.Collections.Generic;

namespace Pulse.CDM
{
  using LungImpairmentMap = Dictionary<eLungCompartment, SEScalar0To1>;
  public class SEPneumoniaExacerbation : SEPatientAction
  {
    protected LungImpairmentMap severities;

    public SEPneumoniaExacerbation()
    {
      severities = new LungImpairmentMap();
    }

    public override void Clear()
    {
      base.Clear();
      foreach (var itr in severities)
        itr.Value.Invalidate();
    }

    public override bool IsValid()
    {
      return HasSeverity();
    }

    public bool HasSeverity()
    {
      foreach (var itr in severities)
        if (itr.Value.IsValid())
          return true;
      return false;
    }
    public bool HasSeverity(eLungCompartment c)
    {
      if (!severities.ContainsKey(c))
        return false;
      return severities[c].IsValid();
    }
    public SEScalar0To1 GetSeverity(eLungCompartment c)
    {
      SEScalar0To1 s = severities[c];
      if (s == null)
      {
        s = new SEScalar0To1();
        severities[c] = s;
      }
      return s;
    }
    public LungImpairmentMap GetSeverities()
    {
      return severities;
    }
  }
}
